services/wfm/src/main/java/org/openkilda/wfm/share/hubandspoke/WorkerBolt.java

```diff
@@ -63,26 +65,38 @@ protected void handleInput(Tuple input) throws AbstractException {
         String sender = input.getSourceComponent();
 
         if (workerConfig.getHubComponent().equals(sender)) {
-            pendingTasks.put(key, input);
+            pendingTasks.put(key, input.getSourceTask());
             registerCallback(key, input);
 
             onHubRequest(input);
-        } else if (pendingTasks.remove(key) != null && workerConfig.getWorkerSpoutComponent().equals(sender)) {
-            cancelCallback(key, input);
+        } else if (pendingTasks.containsKey(key) && workerConfig.getWorkerSpoutComponent().equals(sender)) {
             onAsyncResponse(input);
         } else {
             unhandledInput(input);
         }
     }
 
-    protected abstract void onHubRequest(Tuple input);
+    /**
+     * Send response to hub bolt once all required responses are received.
+     * @param input received tuple.
+     * @param values response to be sent to the hub.
+     */
+    protected void emitResponseToHub(Tuple input, Values values) throws PipelineException {
+        String key = input.getStringByField(MessageTranslator.KEY_FIELD);
+        cancelCallback(key, input);
+
+        int taskId = pendingTasks.remove(key);
```

Comment:
Should we recheck existence of this key into `pendingTasks`?

Reply:
done